substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/SerializationSupport.java

```diff
@@ -0,0 +1,199 @@
+                reportError((FeatureImpl.BeforeAnalysisAccessImpl) access, MULTIPLE_CHECKSUMS, sb.toString());
+            }
+        } catch (Exception e) {
+            throw VMError.shouldNotReachHere(e);
+        }
+
+        return buildTimeConsClass;
+    }
+
+    @Override
+    public Object getSerializationConstructorAccessorClass(Class<?> serializationTargetClass, String targetConstructorClass) {
+        String className = serializationTargetClass.getName();
+        CachedEntity ret = cachedSerializationConstructorAccessors.get(className);
+        if (ret == null) {
+            throw VMError.unsupportedFeature("SerializationConstructorAccessor class is not found for class :" + className + "\n" +
+                            "Generating SerializationConstructorAccessor classes at runtime is not supported. ");
+        } else {
+            Object accessor = ret.serializationConstructorAccessor;
+            Long configuredChecksum = ret.configuredChecksum;
+            long runtimeChecksum = getSerializationChecksum(serializationTargetClass, targetConstructorClass);
+            // configuredChecksum could be null if it is not set in the configuration.
+            if ((configuredChecksum != null && configuredChecksum.longValue() == runtimeChecksum) || configuredChecksum == null) {
+                return accessor;
+            } else {
+                throw VMError.unimplemented("Serialization target class " + className + "'s hierarchy has been changed at run time. Configured checksum is " + configuredChecksum +
+                                ", runtime checksum is " + runtimeChecksum);
+            }
+        }
+    }
+
+    private long getSerializationChecksum(Class<?> serializationTargetClass, String constructorClassName) {
+        if (constructorClassName != null && constructorClassName.length() > 0) {
+            long checksum = 0L;
+            String currentClassName = serializationTargetClass.getName();
+            Class<?> currentClass = serializationTargetClass;
+            // Compute SUID for each serializable superclass
+            while (!constructorClassName.equals(currentClassName)) {
+                Method computeDefaultSUID = ReflectionUtil.lookupMethod(ObjectStreamClass.class, "computeDefaultSUID", Class.class);
+                try {
+                    long classSUID = (Long) computeDefaultSUID.invoke(null, currentClass);
+                    checksum = checksum * 31 + classSUID;
+                    currentClass = currentClass.getSuperclass();
+                    currentClassName = currentClass.getName();
+                } catch (InvocationTargetException | IllegalAccessException e) {
+                    throw VMError.shouldNotReachHere(e);
+                }
+            }
+
+            // Combine checksum with first unserializable superclass
+            return checksum * 31 + constructorClassName.hashCode();
+        } else {
+            return 0L;
+        }
+    }
```

Comment:
Please try extracting the common algorithm in this method and in `BreakpointInterceptor.getChecksumFromOSC` into a method that can be called from both places with values from the different sources.

Reply:
I found it's difficult to apply a common method for Java and JVMTI Agent code because `Object` and `Word` are two incompatible systems. I tried to extract the algorithm into the following code:
```java
public class SerializationChecksumCalculator<T> {
    public long calculate(String targetConstructorClassName,
                                String serializationClassName,
                                T serializationClass,
                                Function<T, Long> calculateFromComputeDefaultSUID,
                                Function<T, String> getClassName,
                                Function<T, T> getSuperClass
    ) {
        long checksum = 0L;
        if (targetConstructorClassName != null && targetConstructorClassName.length() > 0) {
            String currentClassName = serializationClassName;
            T currentClass = serializationClass;
            while (!targetConstructorClassName.equals(currentClassName)) {
                long classSUID = calculateFromComputeDefaultSUID.apply(currentClass);
                checksum = checksum * 31 + classSUID;
                currentClass = getSuperClass.apply(currentClass);
                currentClassName = getClassName.apply(currentClass);
            }
            checksum = checksum * 31 + targetConstructorClassName.hashCode();
        }
        return checksum;
    }
}
``` 
`T` is supposed to be `Class<?>` when it is called from `SerializationSupport`, and to be `JNIObjectHandler` when it is called from `BreakpointInterceptor`. But in fact `JNIObjectHandler` is `Word` based and incompatible with `Object`.

Reply:
I see. Perhaps we can put the two methods in the same class then, so that it is more obvious that both must be changed, or at least add a comment to that effect.

Reply:
> Looks good now! Please attend to the gate checks that are still failing.

I got checkstyle errors on Alibaba group copy write declaration in Java file headers such as 
`/home/runner/work/graal/graal/substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/restrict/SerializationAccessVerifier.java:3: Line does not match expected header line of ' \* DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER\.'.`
How can I fix this kind error?